### PR TITLE
Add Semantic Versioning and refuses outdated Software from Startup

### PR DESCRIPTION
**Pull request by stefanbenten**
This PR adds Versioning Information to Satellite, Storagenode and Bootstrap. It defines also a new VersionedClient that handles GPRC Connections and ensures, the client side software is always up to date.

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Could the PR be broken into smaller PRs?
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?

### CODE REVIEW

File: cmd/storagenode/dashboard.go [hunk 1]
```diff
@@ -80,7 +81,7 @@ func printDashboard(data *pb.DashboardResponse) error {
 	color.NoColor = !useColor
 
 	heading := color.New(color.FgGreen, color.Bold)
-	_, _ = heading.Printf("\nStorage Node Dashboard\n")
+	_, _ = heading.Printf("\nStorage Node Dashboard ( Node Version: %s)\n", version.Build.Version.String())
```
Comment (egonelbre) [inline cmd/storagenode/dashboard.go]: This should come from the dashboard request.

File: internal/version/version.go [hunk 2]
```diff
@@ -5,44 +5,175 @@ package version
...
+	CommitHash string
+	// Version is the semantic version set at compilation
+	// if not a valid semantic version Release should be false
+	Version = "v0.1.0"
+	// Release indicates whether the binary compiled is a release candidate
+	Release bool
+	// Build is a struct containing all relevant build information associated with the binary
+	Build Info
+	// Allowed ensures, the client is still on the allowed versions returned by the control server
+	Allowed bool
```
Comment (egonelbre) [inline internal/version/version.go]: This shouldn't be needed

File: internal/version/version.go [hunk 3]
```diff
@@ -5,44 +5,175 @@ package version
...
+	for _, n := range all {
+		if x == n {
+			return true
+		}
+	}
+	return false
+}
+
+// StrListToSemVerList converts a list of versions to a list of SemVer
+func StrListToSemVerList(serviceVersions []string) (versions []SemVer, err error) {
```
Comment (egonelbre) [inline internal/version/version.go]: StringsToSemVerList

File: pkg/process/exec_conf.go [hunk 4]
```diff
@@ -48,6 +51,8 @@ var (
 
 	contextMtx sync.Mutex
 	contexts   = map[*cobra.Command]context.Context{}
+
+	verClient *version.Client
```
Comment (egonelbre) [inline pkg/process/exec_conf.go]: This shouldn't be a global variable

File: internal/version/service.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+// LogAndReportVersion logs the current version information
+// and reports to monkit
+func (cl *Client) LogAndReportVersion(ctx context.Context) (err error) {
+	err = cl.checkVersionStartup(&ctx)
+	if err != nil {
+		return err
+	}
+
+	//Start up periodic checks
+	go func(ctx context.Context) {
```
Comment (egonelbre) [inline internal/version/service.go]: Use sync2.Cycle here, it allows to manually control the loop from tests.

File: pkg/process/exec.go [hunk 6]
```diff
@@ -9,6 +9,9 @@ import (
 	"os"
 
 	"github.com/spf13/cobra"
+
+	// We use a blank import here to get the side effects from the init function in version
+	_ "storj.io/storj/internal/version"
```
Comment (egonelbre) [inline pkg/process/exec.go]: Shouldn't be necessary.

File: pkg/process/debug.go [hunk 7]
```diff
@@ -32,6 +32,8 @@ func initDebug(logger *zap.Logger, r *monkit.Registry) (err error) {
 	mux.HandleFunc("/debug/pprof/profile", pprof.Profile)
 	mux.HandleFunc("/debug/pprof/symbol", pprof.Symbol)
 	mux.HandleFunc("/debug/pprof/trace", pprof.Trace)
+	mux.HandleFunc("/debug/version", verClient.DebugHandler)
```
Comment (egonelbre) [inline pkg/process/debug.go]: Shouldn't be necessary since each peer creates their own version service.

File: pkg/process/exec_conf.go [hunk 8]
```diff
@@ -237,6 +242,24 @@ func cleanup(cmd *cobra.Command) {
 			delete(contexts, cmd)
 			contextMtx.Unlock()
 		}()
+		//ToDo: Make Server Address a config value
+
+		verClient = &version.Client{
```
Comment (egonelbre) [inline pkg/process/exec_conf.go]: This should be inside peers now.

File: bootstrap/peer.go [hunk 9]
```diff
@@ -79,11 +82,12 @@ type Peer struct {
 }
 
 // New creates a new Bootstrap Node.
-func New(log *zap.Logger, full *identity.FullIdentity, db DB, config Config) (*Peer, error) {
+func New(log *zap.Logger, full *identity.FullIdentity, verClnt *version.Client, db DB, config Config) (*Peer, error) {
```
Comment (egonelbre) [inline bootstrap/peer.go]: Argument should be version info not the service itself.

File: bootstrap/peer.go [hunk 10]
```diff
@@ -79,11 +82,12 @@ type Peer struct {
...
 	peer := &Peer{
 		Log:      log,
 		Identity: full,
 		DB:       db,
+		Version:  verClnt,
 	}
 
 	var err error
```
Comment (egonelbre) [inline bootstrap/peer.go]: The version service should be created somewhere here.

File: bootstrap/peer.go [hunk 11]
```diff
@@ -174,6 +178,9 @@ func New(log *zap.Logger, full *identity.FullIdentity, db DB, config Config) (*P
 func (peer *Peer) Run(ctx context.Context) error {
 	group, ctx := errgroup.WithContext(ctx)
 
+	group.Go(func() error {
+		return ignoreCancel(peer.Version.LogAndReportVersion(ctx))
```
Comment (egonelbre) [inline bootstrap/peer.go]: For consitency name in `Run`

File: cmd/bootstrap/main.go [hunk 12]
```diff
@@ -94,7 +95,9 @@ func cmdRun(cmd *cobra.Command, args []string) (err error) {
 		return errs.New("Error creating tables for master database on bootstrap: %+v", err)
 	}
 
-	peer, err := bootstrap.New(log, identity, db, runCfg)
+	verClient := version.NewClient()
+
+	peer, err := bootstrap.New(log, identity, verClient, db, runCfg)
```
Comment (egonelbre) [inline cmd/bootstrap/main.go]: `version.CompileInfo()` as argument

File: cmd/identity/main.go [hunk 13]
```diff
@@ -168,7 +169,9 @@ func cmdAuthorize(cmd *cobra.Command, args []string) error {
 		config.Signer.Address = defaultSignerAddress
 	}
 
-	signedChainBytes, err := config.Signer.Sign(ctx, ident, authToken)
+	verClient := version.NewClient()
+
+	signedChainBytes, err := config.Signer.Sign(ctx, ident, verClient, authToken)
```
Comment (egonelbre) [inline cmd/identity/main.go]: Passing version info the the sign function feels wrong.

File: internal/testplanet/planet.go [hunk 14]
```diff
@@ -472,7 +481,9 @@ func (planet *Planet) newSatellites(count int) ([]*satellite.Peer, error) {
 		config.Console.StaticDir = filepath.Join(storjRoot, "web/satellite")
 		config.Mail.TemplatePath = filepath.Join(storjRoot, "web/satellite/static/emails")
 
-		peer, err := satellite.New(log, identity, db, &config)
+		verClient := planet.NewVersionClient()
```
Comment (egonelbre) [inline internal/testplanet/planet.go]: Service should be created inside the peer and the version control address passed through the config.

File: internal/version/service.go [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+)
+
+// Client contains the necessary Information to check the Software Version
+type Client struct {
```
Comment (egonelbre) [inline internal/version/service.go]: This should be `Config`, with `Allowed` field removed.

File: internal/version/service.go [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+	ServerAddress  string
+	RequestTimeout time.Duration
+	CheckInterval  time.Duration
+	Allowed        bool
+}
+
+// NewClient creates a Version Check Client with default configuration
+func NewClient() (client *Client) {
```
Comment (egonelbre) [inline internal/version/service.go]: `NewService() *Service`

File: internal/version/version.go [hunk 17]
```diff
@@ -5,44 +5,173 @@ package version
...
 
+// SemVer represents a semantic version
+type SemVer struct {
+	Major int64 `json:"major"`
+	Minor int64 `json:"minor"`
+	Patch int64 `json:"patch"`
+}
+
+// Versions provides a list of SemVer per Service
+type Versions struct {
```
Comment (egonelbre) [inline internal/version/version.go]: Maybe `AllowedVersions` or `Allowed` something like that?

File: cmd/identity/main.go (hunk 13 above)
Comment (stefanbenten) [reply cmd/identity/main.go]: Yeah, it kind of is.
But it is using the transport client to make a GRPC Call..

Comment (egonelbre) [reply cmd/identity/main.go]: Maybe the argument should be the transport.Client instead?

File: internal/version/service.go [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+	/*if !service.IsUpToDate() {
+		zap.S().Fatal("Software Version outdated, please update")
+	}*/
+	return transport
+}
+
+// Run logs the current version information
+func (srv *Service) Run(ctx context.Context) error {
+	return srv.Loop.Run(ctx, func(ctx context.Context) error {
+		srv.mu.Lock()
```
Comment (egonelbre) [inline internal/version/service.go]: Only take for the duration when changing srv.allowed

File: internal/version/service.go [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+func (srv *Service) IsUpToDate() bool {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+	return srv.allowed
+}
+
+// CheckVersion checks if the client is running latest/allowed code
+func (srv *Service) checkVersion(ctx *context.Context) (allowed bool, err error) {
+	defer mon.Task()(ctx)(&err)
+	accepted, err := srv.queryVersionFromControlServer()
```
Comment (egonelbre) [inline internal/version/service.go]: this should take ctx as param

**Comment by stefanbenten** [comment]
@egonelbre Not sure where these data races are coming from..
My tests locally pass just fine.

File: pkg/process/exec.go (hunk 6 above)
Comment (stefanbenten) [reply pkg/process/exec.go]: Thats used to verify the Version Info on build time, that were handed over with the ldfags.
I am not sure what the best way is to do that.

File: cmd/storagenode/dashboard.go (hunk 1 above)
Comment (stefanbenten) [reply cmd/storagenode/dashboard.go]: True, my plan is to add that in addition to the KAD Protobuf changes!

File: cmd/release/main.go [hunk 20]
```diff
@@ -0,0 +1,51 @@
+// Copyright (C) 2019 Storj Labs, Inc.
```
Comment (jtolds) [inline cmd/release/main.go]: i think i'd still like to replace this file with the bash script from the other PR

Comment (stefanbenten) [reply cmd/release/main.go]: Oh yeah. I wanted lay an outline for a future smaller PR.
Should i remove them for now entirely?

Comment (jtolds) [reply cmd/release/main.go]: yeah

**Comment by jtolds** [comment]
merging one sb/ branch into another!

**Comment by stefanbenten** [comment]
https://github.com/storj/storj/pull/1576 is the main PR for getting all of this merged into master 👍

**Pull request merged**